### PR TITLE
feat(connector): [Adyen] implement Kakao for Adyen

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Kakao pay is a South Korean specific payment method.

```
"payment_method_type": "kakao_pay",
  "payment_method_data": {
    "wallet": {
      "kakao_pay_redirect": {}
    }
  }
```

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Tested it using hs-payments-test.netlify.app
Screen Shot 2023-06-28 at 11 44 48 AM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/tests/connectors/adyen_uk_ui.rs [hunk 1]
```diff
@@ -84,6 +84,29 @@ async fn should_make_adyen_klarna_mandate_payment(c: WebDriver) -> Result<(), We
     Ok(())
 }
 
+async fn should_make_adyen_alipay_hk_payment(c: WebDriver) -> Result<(), WebDriverError> {
```
Comment (SanchithHegde) [inline crates/router/tests/connectors/adyen_uk_ui.rs:87]: Please avoid using single character variable names.

File: crates/router/src/connector/adyen/transformers.rs [hunk 2]
```diff
@@ -604,6 +613,24 @@ pub struct AliPayData {
     payment_type: PaymentType,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct AliPayHkData {
+    #[serde(rename = "type")]
+    payment_type: PaymentType,
+}
```
Comment (SanchithHegde) [inline crates/router/src/connector/adyen/transformers.rs:620]: How does the request sent to Adyen look like? I see a `#[serde(tag = "type")]` annotation on the `AdyenPaymentMethod` type, and a `#[serde(rename = "type")]` annotation here. Aren't they kinda redundant?

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/router/src/connector/adyen/transformers.rs [hunk 3]
```diff
@@ -1503,9 +1569,10 @@ impl<'a> TryFrom<(&types::PaymentsAuthorizeRouterData, &api::WalletData)>
         let additional_data = get_additional_data(item);
         let payment_method = AdyenPaymentMethod::try_from(wallet_data)?;
         let shopper_interaction = AdyenShopperInteraction::from(item);
+        let channel = get_channel_type(&item.request.payment_method_type);
         let (recurring_processing_model, store_payment_method, shopper_reference) =
             get_recurring_processing_model(item)?;
-        let return_url = item.request.get_return_url()?;
+        let return_url = item.request.get_router_return_url()?;
```
Comment (jagan-jaya) [inline crates/router/src/connector/adyen/transformers.rs:1575]: remove get_return_url() from the connector utils. if that's a big change create a github issue and we will pick it u later

Comment (AkshayaFoiger) [reply crates/router/src/connector/adyen/transformers.rs:1575]: okay

File: crates/router/src/connector/adyen/transformers.rs (hunk 2 above)
Comment (AkshayaFoiger) [reply crates/router/src/connector/adyen/transformers.rs]: Yes `#[serde(rename = "type")]` is a redundant annotation. But all the payment methods implemented for Adyen uses it. Can I raise a seperate PR to fix it along with other general fix

**Review by SanchithHegde — APPROVED** [review]

File: crates/router/src/connector/adyen/transformers.rs (hunk 2 above)
Comment (SanchithHegde) [reply crates/router/src/connector/adyen/transformers.rs]: Yeah, it should be fine to take this up in a separate PR.

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/src/connector/adyen/transformers.rs [hunk 4]
```diff
@@ -269,6 +269,7 @@ pub struct Amount {
 
 #[derive(Debug, Clone, Serialize)]
 #[serde(tag = "type")]
+#[serde(rename_all = "snake_case")]
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:272]: This is not required

File: crates/router/src/connector/adyen/transformers.rs [hunk 5]
```diff
@@ -1803,20 +1813,21 @@ pub fn get_adyen_response(
...
-            message: response
-                .refusal_reason
-                .unwrap_or_else(|| consts::NO_ERROR_MESSAGE.to_string()),
-            reason: None,
-            status_code,
-        })
-    } else {
-        None
-    };
+    let error: Option<types::ErrorResponse> =
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:1816]: This change is not required

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Pull request merged**